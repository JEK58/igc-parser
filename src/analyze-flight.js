const turf = require('@turf/turf');

function analyzeFlight(flight, task) {
  let nextTP = 0;

  let start, finish;

  for (let i = 0; i < flight.length - 1; i++) {
    let fix1 = flight[i].coordinate;
    let fix2 = flight[i + 1].coordinate;

    if (nextTP < 2) {
      let point = task.points[0].observationZone.checkEnter(fix1, fix2);
      if (point) {
        nextTP = 1;
        start = { point, i, secOfDay: flight[i].secOfDay };
      }
    }

    if (nextTP === task.points.length - 1) {
      let point = task.points[nextTP].observationZone.checkEnter(fix1, fix2);
      if (point) {
        nextTP += 1;
        finish = { point, i, secOfDay: flight[i].secOfDay };
        break;
      }
    }

    let point = task.points[nextTP].observationZone.checkEnter(fix1, fix2);
    if (point) {
      nextTP += 1;
    }
  }

  return { start, finish };
}

module.exports = analyzeFlight;
