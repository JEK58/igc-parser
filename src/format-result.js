function formatResult(result) {
  return `Start: ${formatTime(result.start.secOfDay)}\nTime: ${formatTime(result.totalTime)}`;
}

function formatTime(secOfDay) {
  let sec = secOfDay % 60;
  let min = ((secOfDay - sec) / 60) % 60;
  let hour = (secOfDay - sec - min * 60) / 3600;
  return `${hour}:${min}:${sec}`;
}

module.exports = formatResult;
