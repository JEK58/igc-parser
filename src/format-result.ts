export function formatResult(result) {
  return `Start: ${formatTime(result.start.secOfDay)}\n` +
    `Time: ${formatTime(result.totalTime)}\n` +
    `Distance: ${(result.distance)} km\n` +
    `Speed: ${(result.speed)} km/h\n`;
}

export function formatTime(secOfDay) {
  let sec = secOfDay % 60;
  let min = ((secOfDay - sec) / 60) % 60;
  let hour = (secOfDay - sec - min * 60) / 3600;
  return `${hour}:${min}:${sec}`;
}
