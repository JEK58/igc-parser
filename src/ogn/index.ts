import {Socket} from "net";
import * as readline from "readline";

export default class OGNClient {
  readonly host = 'aprs.glidernet.org';
  readonly port = 10152;

  readonly user = 'test';
  readonly pass = '-1';
  readonly appName = 'test';
  readonly appVersion = '1.0';

  readonly senders: string[];

  private readonly socket = new Socket();
  private readonly reader = readline.createInterface({ input: this.socket as NodeJS.ReadableStream });

  constructor(senders: string[]) {
    this.senders = senders;
    this.reader.on('line', line => this.handleLine(line));
  }

  connect() {
    this.socket.connect(this.port, this.host, () => {
      this.socket.write(`user ${this.user} pass ${this.pass} vers ${this.appName} ${this.appVersion}\n`);
      this.socket.emit('ready');
    });
  }

  handleLine(line: string) {
    let knownSender = this.senders.some(sender => line.indexOf(sender) !== -1);
    if (knownSender) {
      this.socket.emit('record', line);
    }
  }

  on(event: string, handler: Function) {
    if (event === 'line') {
      this.reader.on.call(this.reader, event, handler);
    } else {
      this.socket.on.call(this.socket, event, handler);
    }
  }
}
