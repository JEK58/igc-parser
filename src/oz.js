const turf = require('@turf/turf');

class Cylinder {
  constructor(center, radius) {
    this.center = center;
    this.radius = radius;
  }
}

class Line {
  constructor(center, length) {
    this.center = center;
    this.length = length;
  }

  update() {
    let p1 = turf.destination(this.center, this.length / 2000, this.bearing + 90);
    let p2 = turf.destination(this.center, this.length / 2000, this.bearing - 90);

    this.coordinates = [p1.geometry.coordinates, p2.geometry.coordinates];
  }
}

module.exports = { Cylinder, Line };
