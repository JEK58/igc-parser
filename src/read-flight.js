const fs = require('fs');

function readFlight(path) {
  return fs.readFileSync(path, 'utf8')
    .split('\n')
    .filter(line => line[0] === 'B')
    .map(line => line.match(/^B(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{3})([NS])(\d{3})(\d{2})(\d{3})([EW])/))
    .filter(Boolean)
    .map(match => {
      let secOfDay = parseInt(match[1]) * 3600 + parseInt(match[2]) * 60 + parseInt(match[3]);
      let lat = parseInt(match[4]) + parseInt(match[5]) / 60 + parseInt(match[6]) / 60000;
      let lon = parseInt(match[8]) + parseInt(match[9]) / 60 + parseInt(match[10]) / 60000;
      let coordinate = [lon, lat];
      return { secOfDay, coordinate };
    });
}

module.exports = readFlight;
