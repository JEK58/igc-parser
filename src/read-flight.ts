import * as fs from "fs";

const RE_HFDTE = /^HFDTE(\d{2})(\d{2})(\d{2})/;
const RE_B = /^B(\d{2})(\d{2})(\d{2})(\d{2})(\d{2})(\d{3})([NS])(\d{3})(\d{2})(\d{3})([EW])/;

export interface Fix {
  secOfDay: number,
  time: number,
  coordinate: GeoJSON.Position,
}

export function readFlight(path: string): Fix[] {
  let lines = fs.readFileSync(path, 'utf8').split('\n');

  let date: number | undefined;
  let fixes: Fix[] = [];

  lines.forEach(line => {
    if (date === undefined) {
      date = findDate(line);
    }

    // TODO handle UTC-midnight wraparound
    if (date) {
      let fix = convertLine(line, date);
      if (fix) {
        fixes.push(fix);
      }
    }
  });

  return fixes;
}

function findDate(line: string): number | undefined {
  let match = line.match(RE_HFDTE);
  if (match) {
    let year = parseInt(match[3]) + 2000;
    let month = parseInt(match[2]) - 1;
    let day = parseInt(match[1]);
    return Date.UTC(year, month, day);
  }
}

function convertLine(line: string, date: number): Fix | undefined {
  let match = line.match(RE_B);
  if (!match) return;

  let secOfDay = parseInt(match[1]) * 3600 + parseInt(match[2]) * 60 + parseInt(match[3]);
  let time = date + secOfDay * 1000;

  let lat = parseInt(match[4]) + parseInt(match[5]) / 60 + parseInt(match[6]) / 60000;
  let lon = parseInt(match[8]) + parseInt(match[9]) / 60 + parseInt(match[10]) / 60000;
  let coordinate = [lon, lat];

  return { secOfDay, time, coordinate };
}
