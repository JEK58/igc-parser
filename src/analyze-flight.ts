import {Fix} from "./read-flight";
import Task from "./task/task";
import RacingTaskSolver from "./task/solver/racing-task-solver";
import AreaTaskSolver from "./task/solver/area-task-solver";

export function analyzeFlight(flight: Fix[], task: Task) {
  if (task.options.isAAT) {
    let analyzer = new AreaTaskSolver(task);
    analyzer.consume(flight);
    return analyzer.result;
  } else {
    let analyzer = new RacingTaskSolver(task);
    analyzer.consume(flight);
    return analyzer.result;
  }
}
