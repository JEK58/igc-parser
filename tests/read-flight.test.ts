import {readFlight} from "../src/read-flight";

describe('readFlight()', () => {
  it(`reads "SW_77flqgg1" flight log correctly`, () => {
    let flight = readFlight(`${__dirname}/../fixtures/2017-07-15-lev/SW_77flqgg1.IGC`);
    expect(flight[0]).toMatchSnapshot();
    expect(flight[1234]).toMatchSnapshot();
    expect(flight[3333]).toMatchSnapshot();
  })
});
