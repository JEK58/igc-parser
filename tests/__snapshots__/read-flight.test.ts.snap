// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readFlight() reads "SW_77flqgg1" flight log correctly 1`] = `
Object {
  "coordinate": Array [
    7.01,
    51.01108333333333,
  ],
  "time": 1500111903000,
  "valid": true,
}
`;

exports[`readFlight() reads "SW_77flqgg1" flight log correctly 2`] = `
Object {
  "coordinate": Array [
    6.52835,
    50.866150000000005,
  ],
  "time": 1500118283000,
  "valid": true,
}
`;

exports[`readFlight() reads "SW_77flqgg1" flight log correctly 3`] = `
Object {
  "coordinate": Array [
    7.0655833333333335,
    51.09908333333333,
  ],
  "time": 1500128791000,
  "valid": true,
}
`;
